operator docs (#25682)
MystenLabs/sui: docs/content/guides/operator/archival-stack-setup.mdx

The storage footprint estimations below are based on the network as of early March 2026. These numbers are directional and will grow with network activity.

**Mainnet (full history from genesis):**

| Table | Size |
|-------|------|
| `transactions` | 8.1 TB |
| `objects` | 4.4 TB |
| `checkpoints` | 832 GB |
| `checkpoints_by_digest` | 15 GB |
| `watermark_alt` | 1.5 GB |
| `epochs` | 83 MB |
| `packages` | TBD (not yet backfilled) |
| `packages_by_id` | TBD (not yet backfilled) |
| `packages_by_checkpoint` | TBD (not yet backfilled) |
| `protocol_configs` | TBD (not yet backfilled) |
| `system_packages` | TBD (not yet backfilled) |
| **Total** | **~13.3 TB** |

**Testnet (full history from genesis):**

| Table | Size |
|-------|------|
| `transactions` | 2.5 TB |
| `objects` | 1.2 TB |
| `checkpoints` | 514 GB |
| `checkpoints_by_digest` | 18 GB |
| `watermark_alt` | 12 MB |
| `epochs` | 97 MB |
| `packages` | TBD (not yet backfilled) |
| `packages_by_id` | TBD (not yet backfilled) |
| `packages_by_checkpoint` | TBD (not yet backfilled) |
| `protocol_configs` | TBD (not yet backfilled) |
| `system_packages` | TBD (not yet backfilled) |
| **Total** | **~4.3 TB** |

> The `transactions` and `objects` tables dominate storage. The `packages`, `packages_by_id`, `packages_by_checkpoint`, `protocol_configs`, and `system_packages` pipelines are new and have not yet been backfilled — their storage contribution will increase once backfill is completed. Growth rate scales with network transaction volume.

Reviewer: ```suggestion
The `transactions` and `objects` tables dominate storage. The `packages`, `packages_by_id`, `packages_by_checkpoint`, `protocol_configs`, and `system_packages` pipelines are new and have not yet been backfilled. Their storage contribution will increase once backfill is completed. Growth rate scales with network transaction volume.
```

| `--write-legacy-data` | No | Do not set this. Enable writing deprecated data formats. |
| `--first-checkpoint` | No | Checkpoint to start indexing from. Defaults to 0 (genesis). |
| `--last-checkpoint` | No | Checkpoint to stop indexing at. Useful for bounded backfill jobs. |
| `--metrics-address` | No | Prometheus metrics bind address. Default: `0.0.0.0:9184`. |

#### Example (backfill from GCS bucket)

```sh
sui-kvstore-alt \
    --config kvstore.toml \
    --chain mainnet \
    my-bigtable-instance \
    --remote-store-gcs mysten-mainnet-checkpoints
```

#### Example (steady-state from fullnode)

```sh
sui-kvstore-alt \
    --chain mainnet \
    my-bigtable-instance \
    --rpc-api-url http://my-fullnode:9000 \
    --streaming-url http://my-fullnode:9000
```

### Indexer configuration {#indexer-configuration}

The indexer accepts an optional TOML configuration file via `--config`. For tip-of-chain indexing, the framework defaults are a good starting point and you can omit the flag entirely. Configuration tuning is primarily needed during [backfill](#backfill-rate-limit-sizing).

#### Top-level options

| Option | Default | Description |
|--------|---------|-------------|

> | `total-max-rows-per-second` | unlimited | Global rate limit shared across all pipelines (rows/sec). Set this during backfill to avoid overwhelming Bigtable. |

Reviewer 2: is there a way an operator can roughly think about the rows/ second in terms of checkpoint data?

> | `max-rows-per-second` | unlimited | Default per-pipeline rate limit (rows/sec). Both this and `total-max-rows-per-second` are enforced, so you can use per-pipeline limits to prevent one pipeline from starving others. In practice this is not been needed and you can omit it. |

Reviewer 2: ```suggestion
| `max-rows-per-second` | unlimited | Default per-pipeline rate limit (rows/sec). Both this and `total-max-rows-per-second` are enforced, so you can use per-pipeline limits to prevent one pipeline from starving others. In practice this has not been needed and you can omit it. |
```

> | `bigtable-connection-pool-size` | 10 | Number of gRPC channels in the Bigtable connection pool. Rule of thumb: 2x your Bigtable node count. |

Reviewer: ```suggestion
| `bigtable-connection-pool-size` | 10 | Number of gRPC channels in the Bigtable connection pool. Recommended: 2x your Bigtable node count. |
```

### Backfill

To index from genesis, start the indexer without `--first-checkpoint`. A full mainnet backfill takes approximately **2–3 days** with a 15-node SSD cluster and a 16 CPU indexer.

For backfill, use a single indexer instance — one instance can generate enough load to fully saturate a 15-node cluster. The 16 CPU / 15-node configuration is the largest scale Mysten Labs has tested. Larger instance types and cluster sizes may work but could hit software bottlenecks that have not been identified yet.

#### Recommended backfill strategy

1. **Scale up** the Bigtable cluster to 15 nodes before starting the backfill.
2. **Run one indexer instance** on a larger machine (16 CPU / 32 GB RAM) with aggressive rate limits (~150,000 rows/sec), using the checkpoint bucket as the data source.
3. Once the backfill catches up to the network tip, **switch to steady-state** (see [Steady-state operation](#steady-state-operation) below).
4. **Scale down** the Bigtable cluster to 5 nodes.

From there, monitor cluster CPU utilization and scale nodes as needed — either with Bigtable autoscaling or by monitoring and scaling manually.

#### Backfill rate limit sizing {#backfill-rate-limit-sizing}

> The rate limit you set in the TOML configuration should be based on your Bigtable cluster size. As a rule of thumb, each Bigtable SSD node can sustain approximately **111 rows/sec per 1% CPU utilization**. Google recommends targeting:

Reviewer: ```suggestion
The rate limit you set in the TOML configuration should be based on your Bigtable cluster size. Each Bigtable SSD node can sustain approximately **111 rows/sec per 1% CPU utilization**. Google recommends targeting:
```
